src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -294,6 +301,82 @@ public void ConnectHeaders_SupplyHeadersThenSupplyMore_Both_Tip_PeerId_Maps_Shou
             Assert.True(peerTipsByPeerIdBefore.Values.Count == peerTipsByPeerIdAfter.Values.Count);
         }
 
+        /// <summary>
+        /// Issue 5 @ Create chained header tree component #1321
+        /// 3 peers should supply headers - one of the peers creates a fork.
+        /// Make sure that ChainedBlock is not created more than once for the same header.
+        /// Check that next pointers were created correctly.
+        /// </summary>
+        [Fact]
+        public void ConnectHeaders_HeadersFromTwoPeersWithFork_ShouldCreateBlocksForNewHeaders()
+        {
+            TestContext testContext = new TestContextBuilder().Build();
+            ChainedHeaderTree chainedHeaderTree = testContext.ChainedHeaderTree;
+            ChainedHeader chainTip = testContext.ExtendAChain(7);
+            chainedHeaderTree.Initialize(chainTip, true);
+            testContext.ChainStateMock.Setup(s => s.ConsensusTip).Returns(chainTip);
+
+            // Peer 1: 1a - 2a - 3a - 4a - 5a - 6a - 7a - 8a
+            Assert.True(chainedHeaderTree.GetChainedHeadersByHash().Count == 8);
+
+            // Peer 2: 1a - 2a - 3a - 4a - 5a - 6a - 7a - 8a
+            List<BlockHeader> listOfExistingHeaders = testContext.ChainedHeaderToList(chainTip, 7);
+            chainedHeaderTree.ConnectNewHeaders(2, listOfExistingHeaders);
+            Assert.True(chainedHeaderTree.GetChainedHeadersByHash().Count == 8);
+
+            // Peer 3: 1a - 2a - 3a - 4a - 5b - 6b - 7b - 8b
+            var forkedBlockHeader = chainTip.Previous.Previous.Previous.Previous;
+            ChainedHeader peerThreeTip = testContext.ExtendAChain(4, forkedBlockHeader);
+            List<BlockHeader> listOfNewAndExistingHeaders = testContext.ChainedHeaderToList(peerThreeTip, 6);   // includes common blocks
+            chainedHeaderTree.ConnectNewHeaders(3, listOfNewAndExistingHeaders);
+
+            // CHT should contain 12 headers
+            Assert.True(chainedHeaderTree.GetChainedHeadersByHash().Count == 12);
+
+            var chainedHeadersByHash = chainedHeaderTree.GetChainedHeadersByHash();
+
+            foreach (int peer in new[] {2, 3})
+            {
+                var nextPointersByHeightMap = new Dictionary<int, HashSet<ChainedHeader>>();
```

Comment:
Instead of Hash set you would be better using list, because the tree was incorrect and contained the same next twice, with Hash set you would not recognise that

Reply:
Changed to List<>.